src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.DeclarationAnalysisData.cs

```diff
@@ -9,14 +10,18 @@ internal abstract partial class AnalyzerDriver : IDisposable
     {
         internal sealed class DeclarationAnalysisData
         {
-            public static DeclarationAnalysisData CreateFrom(DeclarationAnalysisDataBuilder builder) => new DeclarationAnalysisData(builder);
-            private DeclarationAnalysisData(DeclarationAnalysisDataBuilder builder)
+            public DeclarationAnalysisData(
+                SyntaxNode declaringReferenceSyntax,
+                SyntaxNode topmostNodeForAnalysis,
+                ImmutableArray<DeclarationInfo>.Builder declarationsInNodeBuilder,
+                IEnumerable<SyntaxNode> descendantNodesToAnalyze,
```

Comment:
can we make descendantNodesToAnalyze an ArrayBuilder as well to reduce the allocations here.

Reply:
we can use ArrayBuilder for all Builder usage and use ToImmutableAndFree to pool builder. or you can move back to List and use IReadOnlyList